pkg/api/grpcServer.go

```diff
@@ -14,314 +14,204 @@
+		} else {
 			// do the actual work for the message
-			args := dispatchArgs{
-				dState.request,
-				dState.response,
-				*dState.requestIndex,
-				requestBag,
-				responseBag,
-				dState.result,
-			}
-
-			worker(ctx2, args)
+			out = worker(ctx, requestBag, responseBag)
 
-			sendLock.Lock()
-			if err = respTracker.ApplyBag(responseBag, 0, dState.outAttrs); err != nil {
-				*dState.outAttrs = mixerpb.Attributes{}
-				glog.Errorf("Unable to apply response attribute bag, sending blank attributes: %v", err)
+			if dState.result != nil {
+				*dState.result = out
+				out = status.OK
 			}
-			err = s.sendMsg(stream, dState.response)
-			sendLock.Unlock()
-
-			if err != nil {
-				glog.Errorf("Unable to send gRPC response message: %v", err)
-			}
-
-			requestBag.Done()
-			responseBag.Done()
-
-			span.LogFields(log.Object("gRPC response", dState.response))
-			span.Finish()
+		}
+	}
 
-			wg.Done()
-		})
+	if dState.outAttrs != nil {
+		responseBag.ToProto(dState.outAttrs, s.wordMap)
 	}
-}
 
-// Check is the entry point for the external Check method
-func (s *grpcServer) Check(stream mixerpb.Mixer_CheckServer) error {
-	return s.dispatch(stream, "/istio.mixer.v1.Mixer/Check", getCheckState, s.preprocess(s.handleCheck))
-}
+	requestBag.Done()
+	responseBag.Done()
 
-// Report is the entry point for the external Report method
-func (s *grpcServer) Report(stream mixerpb.Mixer_ReportServer) error {
-	return s.dispatch(stream, "/istio.mixer.v1.Mixer/Report", getReportState, s.preprocess(s.handleReport))
+	return makeGRPCError(out)
 }
 
-// Quota is the entry point for the external Quota method
-func (s *grpcServer) Quota(stream mixerpb.Mixer_QuotaServer) error {
-	return s.dispatch(stream, "/istio.mixer.v1.Mixer/Quota", getQuotaState, s.preprocess(s.handleQuota))
+func makeGRPCError(status rpc.Status) error {
+	return grpc.Errorf(codes.Code(status.Code), status.Message)
 }
 
-func (s *grpcServer) handleCheck(ctx context.Context, args dispatchArgs) {
-
-	resp := args.response.(*mixerpb.CheckResponse)
+// Check is the entry point for the external Check method
+func (s *grpcServer) Check(ctx context2.Context, req *mixerpb.CheckRequest) (*mixerpb.CheckResponse, error) {
+	resp := &mixerpb.CheckResponse{}
 
-	if glog.V(2) {
-		glog.Infof("Check [%x]", args.requestIndex)
+	dState := dispatchState{
+		inAttrs:  &req.Attributes,
+		outAttrs: &resp.Attributes,
+		result:   &resp.Status,
 	}
 
-	*args.result = s.aspectDispatcher.Check(ctx, args.requestBag, args.responseBag)
-	// TODO: this value needs to initially come from config, and be modulated by the kind of attribute
-	//       that was used in the check and the in-used aspects (for example, maybe an auth check has a
-	//       30s TTL but a whitelist check has got a 120s TTL)
-	resp.Expiration = 5 * time.Second
+	err := s.dispatch(ctx, &dState, func(ctx context.Context, requestBag *attribute.MutableBag, responseBag *attribute.MutableBag) rpc.Status {
```

Comment:
here we have two contexts of different types both named `ctx`. can we update one or the other to be clearer about what is used later?